packages/flutter/lib/src/cupertino/route.dart

```diff
@@ -387,48 +387,106 @@ class CupertinoPageTransition extends StatelessWidget {
-  }) : _primaryPositionAnimation =
-           (linearTransition
-             ? primaryRouteAnimation
-             : CurvedAnimation(
-                 parent: primaryRouteAnimation,
-                 curve: Curves.fastEaseInToSlowEaseOut,
-                 reverseCurve: Curves.fastEaseInToSlowEaseOut.flipped,
-               )
-           ).drive(_kRightMiddleTween),
-       _secondaryPositionAnimation =
-           (linearTransition
-             ? secondaryRouteAnimation
-             : CurvedAnimation(
-                 parent: secondaryRouteAnimation,
-                 curve: Curves.linearToEaseOut,
-                 reverseCurve: Curves.easeInToLinear,
-               )
-           ).drive(_kMiddleLeftTween),
-       _primaryShadowAnimation =
-           (linearTransition
-             ? primaryRouteAnimation
-             : CurvedAnimation(
-                 parent: primaryRouteAnimation,
-                 curve: Curves.linearToEaseOut,
-               )
-           ).drive(_CupertinoEdgeShadowDecoration.kTween);
+    required this.linearTransition,
+  });
+
+  /// A linear route animation from 0.0 to 1.0 when this screen is being pushed.
+  final Animation<double> primaryRouteAnimation;
+
+  /// A linear route animation from 0.0 to 1.0 when another screen is being
+  /// pushed on top of this one.
+  final Animation<double> secondaryRouteAnimation;
+
+  /// Whether to perform the transitions linearly. Used to precisely track back
+  /// gesture drags.
+  final bool linearTransition;
 
-  // When this page is coming in to cover another page.
-  final Animation<Offset> _primaryPositionAnimation;
-  // When this page is becoming covered by another page.
-  final Animation<Offset> _secondaryPositionAnimation;
-  final Animation<Decoration> _primaryShadowAnimation;
 
   /// The widget below this widget in the tree.
   final Widget child;
 
+  @override
+  State<CupertinoPageTransition> createState() => _CupertinoPageTransitionState();
+}
+
+class _CupertinoPageTransitionState extends State<CupertinoPageTransition> {
+
+  CurvedAnimation? _primaryPositionCurvedAnimation;
+
+  /// When this page is coming in to cover another page.
+  late Animation<Offset> _primaryPositionAnimation;
+
+  CurvedAnimation? _backingSecondaryPositionAnimation;
+
+  /// When this page is becoming covered by another page.
+  late Animation<Offset> _secondaryPositionAnimation;
+
+  CurvedAnimation? _backingPrimaryShadowAnimation;
+  late Animation<Decoration> _primaryShadowAnimation;
+
+  @override
+  void initState()  {
+    super.initState();
+    _setAnimations();
+  }
+
+  @override
+  void didUpdateWidget(CupertinoPageTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (
+      oldWidget.linearTransition != widget.linearTransition ||
+      oldWidget.primaryRouteAnimation != widget.primaryRouteAnimation ||
+      oldWidget.secondaryRouteAnimation != widget.secondaryRouteAnimation) {
+      _setAnimations();
+    }
+  }
```

Comment:
@polina-c I'm having some trouble with some tests. (ex: `packages/flutter/test/cupertino/route_test.dart` "Animated push/pop is not linear").

https://github.com/flutter/flutter/blob/98685a099fdde0b0424b5665debbe508a5c697e5/packages/flutter/test/cupertino/route_test.dart#L928

Looking at the values (the `.dx`s) of the test, it looks like they are correct.

However, with my changes, it fails here:

https://github.com/flutter/flutter/blob/98685a099fdde0b0424b5665debbe508a5c697e5/packages/flutter/test/cupertino/route_test.dart#L965-L968

```console
The following TestFailure was thrown running a test:
Expected: -197.0 (±1.0)
  Actual: <-265.3022323052088>
   Which: -265.3022323052088 is not in the range of -197.0 (±1.0).
```

It looks like, with my changes, it is not using the `reverseCurve` given to the `CurvedAnimation` anymore.

Interestingly, the test passes if I do 

```suggestion
  @override
  void didUpdateWidget(CupertinoPageTransition oldWidget) {
    super.didUpdateWidget(oldWidget);
    _setAnimations();
  }
```

Do you have an idea of what could be going on?